README.md
# turtle.io

[![build status](https://secure.travis-ci.org/avoidwork/turtle.io.svg)](http://travis-ci.org/avoidwork/turtle.io) [![Gitter](https://badges.gitter.im/Join Chat.svg)](https://gitter.im/avoidwork/turtle.io?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

turtle.io is an HTTP server that gets faster by learning as it handles traffic. It achieves > 99% concurrency with a higher amount of HTTP transactions/s than other popular servers.

turtle.io is very easy to get up and running! All you need to do is install it, and tell it what directory holds your web sites, & which hostnames to answer for.

You can also create complex web applications, with a familiar API.

## Getting Started
1. Install the module with: `npm install turtle.io`
2. Create a script to load & start a server. You could use `sample.js` in the turtle.io directory (./node_modules/turtle.io) as a template, or see the examples below
3. [Optional] Edit `config.json` in the turtle.io directory to configure server defaults; you can override defaults by passing server.start() an Object

## Examples
turtle.io requires a ***default*** virtual host to be specified, because it is the failover when a request can't be routed.

#### Virtual hosts
Virtual host keys are the hostname, and the value is the directory relative to "root".

```javascript
var server = require("turtle.io")();
server.start({
	default : "mysite.com",
	port    : 80,
	uid     : 100,
	root    : "/var/www",
	pages   : "errors",
	vhosts  : {
		"mysite.com"         : "mysite.com",
		"another-domain.com" : "another-domain.com"
	}
});
```

#### Proxy routes
This example has `/api` act as a reverse proxy to another service.

```javascript
var server = require("turtle.io")();
server.proxy("/api", "https://api.github.com");
server.start(require(__dirname + "/config.json"));
```

## Benchmark with express.js
#### Specs
- **Machine** MacBook Air (Early '14) / Core i7 @ 1.7Ghz / 8GB ram / 512 flash / OS X 10.10.2
- **ulimit** 2560
- **express** 4.11.2
- **turtle.io** 3.2.2
- **benchmark** ```siege -c100 -b -q -H 'Connection: Keep-Alive' -t15S localhost:$@```

#### Test
##### express.js
`Hello World!` from a route (content-length: 12), no `allow` header.

```javascript
var express = require("express"),
    app = express();

app.get('/', function (req, res) {
  res.send('Hello World!')
})

app.listen(3000);
```

##### turtle.io
`Hello World!` html file streamed from disk (content-length: 53), has accurate `allow` header.

```javascript
"use strict";

var dir    = __dirname,
    server = require(dir + "/lib/turtle.io")();

server.start( {
	default : "test",
	root    : dir + "/sites",
	vhosts  : {
		"test" : "test"
	},
	logs: {
		stdout: false
	}
} );
```
#### Transactions/s
- **turtle.io** ```[1233.78, 1203.2, 1187.44]  (1208.14 avg)```
- **express**   ```[1105.85, 1124.25, 1167.57] (1132.56 avg)```

#### Concurrency
- **turtle.io** ```[99.33, 99.47, 99.57] (99.46 avg)```
- **express**   ```[42.81, 41.2, 44.27]  (42.76 avg)```

## Handling Uploads
The `request` object is passed to every route handler as the second argument, will have a `body` property with the payload from the Client. It will not be coerced to another format, so if you expect JSON, you'll have to `JSON.parse()` it yourself (for now).

## API & decoration
#### request
##### allow
_String_

Allowed HTTP methods

##### ip
_Number_

Request IP

##### parsed
_Object_

Parsed HTTP request

##### query
_String_

Parsed query string

##### server
_Object_

turtle.io instance

##### vhost
_String_

Virtual host handling the request.

#### respond
##### error
_Function (status, body)_

Send an error response.

##### redirect
_Function (url)_

Send a redirection.

##### respond
_Function (body[, status, headers])_

Send a response.

## Configuration
Configuration values can be set by editing `config.json` in the turtle.io directory, or by passing an Object to `start()`.

#### cache
_Number (1000)_

Size of LRU cache for Etag validation.

#### catchAll
_Boolean (true)_

Handle unterminated requests

#### default
_String_

***[Required]*** Default hostname to handle requests which are not specified within _vhosts_; must be a valid entry within _vhosts_.

#### headers
_Object_

Response headers. CORS is enabled by default.

#### id
_String (turtle_io)_

DTrace application identifier

#### index
_Array_

Files to look for when accessing a directory resource.

#### json
_Number (2)_

Default "pretty" ident size

#### logs
_Object_

Logging configuration.

#### logs.format
_String_ (%v %h %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-agent}i\")

Common Log Format string of tokens, defaulting to standard Virtual Host format.

#### logs.level
_String_ ("info")

Minimum Common Log Level which is emitted to `stdout`.

#### logs.stdout
_Boolean_ (true)

Override & disable `stdout` emitting by setting to `false`.

#### logs.dtrace
_Boolean_ (false)

Override & enable `dtrace` probes which emit nanosecond timing of hot paths.

#### logs.time
_String_ (D/MMM/YYYY:HH:mm:ss ZZ)

Format for the date/time portion of a log message.

#### maxBytes
_Number (0/unlimited)_

Default maximum request body size, when exceeded a 429 is sent.

#### pages
_String (null)_

Directory relative to `root` which has files named for HTTP status codes, to be served upon error, e.g. `404.htm`.

#### port
_Number (8000)_

Port the server will listen on.

#### proxy
_Object_

Proxy configuration.

#### proxy.rewrite
_Array (["index.htm", "index.html"])_

Content-Type header values to apply URL rewrites to.

#### root
_String ("")_

Relative path to the web root directory.

#### ssl.cert
_Object_

[Optional] SSL certificate

#### ssl.key
_Object_

[Optional] SSL certificate key/pem

#### uid
_Number (null)_

[Optional] UID the server runs as.

#### vhosts
_Object_

***[Required]*** Virtual hosts the server will respond for, `key` is the hostname & `value` is the directory relative to `root`.

## License
Copyright (c) 2015 Jason Mulligan  
Licensed under the BSD-3 license.


CHANGELOG.md
# Change Log

## [4.0.3](https://github.com/avoidwork/turtle.io/tree/4.0.3) (2015-05-26)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/4.0.2...4.0.3)

**Implemented enhancements:**

- Fix file organization [\#334](https://github.com/avoidwork/turtle.io/issues/334)

## [4.0.2](https://github.com/avoidwork/turtle.io/tree/4.0.2) (2015-05-25)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/4.0.1...4.0.2)

**Closed issues:**

- Reverse proxy regression with self-signed certs [\#332](https://github.com/avoidwork/turtle.io/issues/332)

## [4.0.1](https://github.com/avoidwork/turtle.io/tree/4.0.1) (2015-05-25)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/4.0.0...4.0.1)

## [4.0.0](https://github.com/avoidwork/turtle.io/tree/4.0.0) (2015-05-25)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.18...4.0.0)

**Implemented enhancements:**

- I'd like to see `res.send\(\)` [\#329](https://github.com/avoidwork/turtle.io/issues/329)

**Fixed bugs:**

- Reverse proxy is busted [\#331](https://github.com/avoidwork/turtle.io/issues/331)

## [3.2.18](https://github.com/avoidwork/turtle.io/tree/3.2.18) (2015-04-03)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.17...3.2.18)

## [3.2.17](https://github.com/avoidwork/turtle.io/tree/3.2.17) (2015-03-31)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.16...3.2.17)

## [3.2.16](https://github.com/avoidwork/turtle.io/tree/3.2.16) (2015-03-30)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.15...3.2.16)

## [3.2.15](https://github.com/avoidwork/turtle.io/tree/3.2.15) (2015-03-30)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.14...3.2.15)

## [3.2.14](https://github.com/avoidwork/turtle.io/tree/3.2.14) (2015-03-29)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.13...3.2.14)

## [3.2.13](https://github.com/avoidwork/turtle.io/tree/3.2.13) (2015-03-29)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.12...3.2.13)

## [3.2.12](https://github.com/avoidwork/turtle.io/tree/3.2.12) (2015-03-26)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.11...3.2.12)

## [3.2.11](https://github.com/avoidwork/turtle.io/tree/3.2.11) (2015-03-26)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.10...3.2.11)

## [3.2.10](https://github.com/avoidwork/turtle.io/tree/3.2.10) (2015-03-20)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.9...3.2.10)

## [3.2.9](https://github.com/avoidwork/turtle.io/tree/3.2.9) (2015-03-18)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.8...3.2.9)

## [3.2.8](https://github.com/avoidwork/turtle.io/tree/3.2.8) (2015-03-18)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.7...3.2.8)

## [3.2.7](https://github.com/avoidwork/turtle.io/tree/3.2.7) (2015-03-16)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.6...3.2.7)

## [3.2.6](https://github.com/avoidwork/turtle.io/tree/3.2.6) (2015-03-13)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.5...3.2.6)

## [3.2.5](https://github.com/avoidwork/turtle.io/tree/3.2.5) (2015-03-12)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.4...3.2.5)

## [3.2.4](https://github.com/avoidwork/turtle.io/tree/3.2.4) (2015-02-25)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.3...3.2.4)

**Fixed bugs:**

- Unneeded value comparison in `run\(\)` [\#327](https://github.com/avoidwork/turtle.io/issues/327)

## [3.2.3](https://github.com/avoidwork/turtle.io/tree/3.2.3) (2015-02-13)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.2...3.2.3)

## [3.2.2](https://github.com/avoidwork/turtle.io/tree/3.2.2) (2015-02-13)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.1...3.2.2)

## [3.2.1](https://github.com/avoidwork/turtle.io/tree/3.2.1) (2015-02-12)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.2.0...3.2.1)

## [3.2.0](https://github.com/avoidwork/turtle.io/tree/3.2.0) (2015-02-04)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.1.25...3.2.0)

**Implemented enhancements:**

- Needs FSGI [\#313](https://github.com/avoidwork/turtle.io/issues/313)

- Needs JSGI [\#312](https://github.com/avoidwork/turtle.io/issues/312)

- Needs CoAP [\#286](https://github.com/avoidwork/turtle.io/issues/286)

## [3.1.25](https://github.com/avoidwork/turtle.io/tree/3.1.25) (2015-01-31)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.1.24...3.1.25)

## [3.1.24](https://github.com/avoidwork/turtle.io/tree/3.1.24) (2015-01-25)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.1.23...3.1.24)

## [3.1.23](https://github.com/avoidwork/turtle.io/tree/3.1.23) (2015-01-20)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.1.22...3.1.23)

**Fixed bugs:**

- `X-Age` header resets [\#326](https://github.com/avoidwork/turtle.io/issues/326)

## [3.1.22](https://github.com/avoidwork/turtle.io/tree/3.1.22) (2015-01-15)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.1.21...3.1.22)

## [3.1.21](https://github.com/avoidwork/turtle.io/tree/3.1.21) (2015-01-15)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.1.20...3.1.21)

## [3.1.20](https://github.com/avoidwork/turtle.io/tree/3.1.20) (2015-01-11)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.1.19...3.1.20)

## [3.1.19](https://github.com/avoidwork/turtle.io/tree/3.1.19) (2015-01-09)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.1.18...3.1.19)

**Fixed bugs:**

- `middleware\(\)` is not handling an Error properly [\#325](https://github.com/avoidwork/turtle.io/issues/325)

## [3.1.18](https://github.com/avoidwork/turtle.io/tree/3.1.18) (2014-12-16)

[Full Changelog](https://github.com/avoidwork/turtle.io/compare/3.1.17...3.1.18)


fresh.js
var $         = require( "abaaso" ),
    crypto    = require( "crypto" ),
    fs        = require( "fs" ),
    http      = require( "http" ),
    https     = require( "https" ),
    http_auth = require( "http-auth" ),
    mime      = require( "mime" ),
    moment    = require( "moment" ),
    syslog    = require( "node-syslog" ),
    toobusy   = require( "toobusy" ),
    zlib      = require( "zlib" );

/**
 * TurtleIO
 *
 * @constructor
 */
function TurtleIO () {
	this.cache    = $.lru( 1000 );
	this.config   = {};
	this.handlers = {all: {regex: [], routes: [], hosts: {}}, "delete": {regex: [], routes: [], hosts: {}}, get: {regex: [], routes: [], hosts: {}}, patch: {regex: [], routes: [], hosts: {}}, post: {regex: [], routes: [], hosts: {}}, put: {regex: [], routes: [], hosts: {}}};
	this.server   = null;
	this.vhosts   = [];
};

// Prototype loop
TurtleIO.prototype.constructor = TurtleIO;

/**
 * Sets a DELETE handler
 *
 * @method delete
 * @param  {String}   route RegExp pattern
 * @param  {Function} fn    Handler
 * @param  {String}   host  [Optional] Virtual host, default is `all`
 * @return {Object}         TurtleIO instance
 */
TurtleIO.prototype["delete"] = function ( route, fn, host ) {
	return this.handler( "delete", route, fn, host );
};

/**
 * Error handler
 *
 * @method error
 * @param  {Object} req  Request Object
 * @param  {Object} res  Response Object
 * @return {Object}      TurtleIO instance
 */
TurtleIO.prototype.error = function ( req, res ) {
	return this;
};

/**
 * Sets a GET handler
 *
 * @method delete
 * @param  {String}   route RegExp pattern
 * @param  {Function} fn    Handler
 * @param  {String}   host  [Optional] Virtual host, default is `all`
 * @return {Object}         TurtleIO instance
 */
TurtleIO.prototype.get = function ( route, fn, host ) {
	return this.handler( "get", route, fn, host );
};

/**
 * Sets a PATCH handler
 *
 * @method delete
 * @param  {String}   route RegExp pattern
 * @param  {Function} fn    Handler
 * @param  {String}   host  [Optional] Virtual host, default is `all`
 * @return {Object}         TurtleIO instance
 */
TurtleIO.prototype.patch = function ( route, fn, host ) {
	return this.handler( "path", route, fn, host );
};

/**
 * Sets a POST handler
 *
 * @method delete
 * @param  {String}   route RegExp pattern
 * @param  {Function} fn    Handler
 * @param  {String}   host  [Optional] Virtual host, default is `all`
 * @return {Object}         TurtleIO instance
 */
TurtleIO.prototype.post = function ( route, fn, host ) {
	return this.handler( "post", route, fn, host );
};

/**
 * Sets a PUT handler
 *
 * @method delete
 * @param  {String}   route RegExp pattern
 * @param  {Function} fn    Handler
 * @param  {String}   host  [Optional] Virtual host, default is `all`
 * @return {Object}         TurtleIO instance
 */
TurtleIO.prototype.put = function ( route, fn, host ) {
	return this.handler( "put", route, fn, host );
};

/**
 * Sets a handler
 *
 * @method handler
 * @param  {String}   method HTTP method
 * @param  {String}   route  RegExp pattern
 * @param  {Function} fn     Handler
 * @param  {String}   host   [Optional] Virtual host, default is `all`
 * @return {Object}          TurtleIO instance
 */
TurtleIO.prototype.handler = function ( method, route, fn, host ) {
	host = host || "all";

	if ( !!this.handlers[method].hosts[host] ) {
		this.handlers[method].hosts[host] = {};
	}

	this.handlers[method].routes.push( route );
	this.handlers[method].regex.push( new RegExp( "^" + route + "$" ) );
	this.handlers[method].hosts[host][route] = fn;

	return this;
};

/**
 * Default request handler
 *
 * @method request
 * @param  {Object} req Request Object
 * @param  {Object} res Response Object
 * @return {Object}     TurtleIO instance
 */
TurtleIO.prototype.request = function ( req, res ) {
	this.respond( req, res, "hi!", 200 );

	return this;
};

/**
 * Send a response
 *
 * @method respond
 * @param  {Object} req     Request Object
 * @param  {Object} res     Response Object
 * @param  {Mixed}  body    Primitive or Buffer
 * @param  {Number} status  [Optional] HTTP status, default is `200`
 * @param  {Object} headers [Optional] HTTP headers
 * @return {Object}         TurtleIO instance
 */
TurtleIO.prototype.respond = function ( req, res, body, status, headers ) {
	status  = status  || 200;
	headers = headers || {Allow: "GET, HEAD, OPTIONS", "Content-Type": "text/plain"};

	res.writeHead( status, headers );
	res.end( body );

	return this;
};

/**
 * Restarts the instance
 *
 * @method restart
 * @return {Object} TurtleIO instance
 */
TurtleIO.prototype.restart = function () {
	var config = this.config;

	this.stop().start( config );
};

/**
 * Routes a request to a handler
 *
 * @method route
 * @param  {Object} req Request Object
 * @param  {Object} res Response Object
 * @return {Object}     TurtleIO instance
 */
TurtleIO.prototype.route = function ( req, res ) {
	var self   = this,
	    parsed = $.parse( req.url ),
	    method = req.method.toLowerCase(),
	    host, route;

	// Finding a matching vhost
	this.vhosts.each( function ( i ) {
		if ( i.test( parsed.host ) ) {
			return ! ( host = i.toString().replace( /^\/\^|\$\/$/g, "" ) );
		}
	} );

	if ( !host ) {
		host = this.config["default"];
	}

	// Looking for a match
	this.handlers[method].regex.each( function ( i, idx ) {
		var x = self.handlers[method].routes[idx];

		if ( x in self.handlers[method].hosts[host] || x in self.handlers[method].hosts.all ) {
			route   = i;
			handler = self.handlers[method].hosts[host][x] || self.handlers[method].hosts.all[x];
			return false;
		}
	} );

	// Looking for a match against generic routes
	if ( !route ) {
		this.handlers.all.regex.each( function ( i, idx ) {
			var x = self.handlers.all.routes[idx];

			if ( x in self.handlers.all.hosts[host] || x in self.handlers.all.hosts.all ) {
				route   = i;
				handler = self.handlers.all.hosts[host][x] || self.handlers.all.hosts.all[x];
				return false;
			}
		} );
	}

	if ( handler ) {
		handler( req, res );
	}
	else {
		this.error( req, res );
	}

	return this;
};

/**
 * Starts the instance
 *
 * @method start
 * @param  {Object}   config Configuration
 * @param  {Function} err    Error handler
 * @param  {Function} msg    Message handler
 * @return {Object}          TurtleIO instance
 */
TurtleIO.prototype.start = function ( config, err, msg ) {
	var self = this;

	config = config || {};
	err    = err    || null;
	msg    = msg    || null;

	// Setting configuration
	if ( !config.port ) {
		config.port = 8000;
	}

	if ( !config.ip ) {
		config.ip = "127.0.0.1";
	}

	this.config = config;

	// Setting default routes
	this.host( "all" );

	// Registering virtual hosts
	$.array.cast( config.vhosts, true ).each( function ( i ) {
		self.host( i );
	} );

	// Setting a default GET route
	if ( !this.handlers.get.routes.contains( ".*" ) ) {
		this.get( ".*", function ( req, res ) {
			self.request( req, res );
		}, "all" );
	}

	// Starting server
	if ( this.config.cert !== undefined ) {
		console.log("ssl!");
	}
	else {
		this.server = http.createServer( function ( req, res ) {
			self.route( req, res );
		} ).listen( config.port, config.ip );
	}

	console.log( "Started turtle.io on port " + config.port );

	return this;
};

/**
 * Stops the instance
 *
 * @method stop
 * @return {Object} TurtleIO instance
 */
TurtleIO.prototype.stop = function () {
	var port = this.config.port;

	this.cache    = $.lru( 1000 );
	this.config   = {};
	this.handlers = {all: {regex: [], routes: [], hosts: {}}, "delete": {regex: [], routes: [], hosts: {}}, get: {regex: [], routes: [], hosts: {}}, patch: {regex: [], routes: [], hosts: {}}, post: {regex: [], routes: [], hosts: {}}, put: {regex: [], routes: [], hosts: {}}};
	this.server   = null;
	this.vhosts   = [];

	console.log( "Stopped turtle.io on port " + port );

	return this;
};

/**
 * Registers a virtual host
 *
 * @method host
 * @param  {String} arg Virtual host
 * @return {Object}     TurtleIO instance
 */
TurtleIO.prototype.host = function ( arg ) {
	this.vhosts.push( new RegExp( "^" + arg + "$" ) );

	this.handlers.all.hosts[arg]       = {};
	this.handlers["delete"].hosts[arg] = {};
	this.handlers.get.hosts[arg]       = {};
	this.handlers.patch.hosts[arg]     = {};
	this.handlers.post.hosts[arg]      = {};
	this.handlers.put.hosts[arg]       = {};

	return this;
};

/**
 * Constructs a URL
 *
 * @method url
 * @param  {Object} req Request Object
 * @return {String}     Requested URL
 */
TurtleIO.prototype.url = function ( req ) {
	return "http" + ( this.config.cert !== undefined ? "s" : "" ) + "://" + req.headers.host + req.url;
};

module.exports = TurtleIO;
